packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp

```diff
@@ -134,12 +135,87 @@ register_object_notification_cb(realm_object_t *object, jobject callback) {
     );
 }
 
+inline void jni_check_exception(JNIEnv* jenv = get_env()) {
+    if (jenv->ExceptionCheck()) {
+        jenv->ExceptionDescribe();
+        throw std::runtime_error("An unexpected Error was thrown from Java.");
+    }
+}
+
+class CustomJVMScheduler {
+public:
+    CustomJVMScheduler(jobject dispatchScheduler) {
+        JNIEnv *jenv = get_env();
+        jclass jvm_scheduler_class = jenv->FindClass("io/realm/internal/interop/JVMScheduler");
+        m_notify_method = jenv->GetMethodID(jvm_scheduler_class, "notifyCore", "(JJ)V");
+        m_jvm_dispatch_scheduler = jenv->NewGlobalRef(dispatchScheduler);
+    }
+
+    ~CustomJVMScheduler() {
+        get_env(true)->DeleteGlobalRef(m_jvm_dispatch_scheduler);
+    }
+
+    void notify() override {
+        auto jenv = get_env(true);
+        jni_check_exception(jenv);
+        jenv->CallVoidMethod(m_jvm_dispatch_scheduler, m_notify_method,
+                             reinterpret_cast<jlong>(&m_callback));
+    }
+
+    void set_notify_callback(std::function<void()> fn) override {
+        m_callback = std::move(fn);
+    }
+
+    bool is_on_thread() const noexcept override {
+        return m_id == std::this_thread::get_id();
+    }
+
+    bool is_same_as(const Scheduler *other) const noexcept override {
+        auto o = dynamic_cast<const CustomJVMScheduler *>(other);
+        return (o && (o->m_id == m_id));
+    }
+
+    bool can_deliver_notifications() const noexcept override {
+        return true;
+    }
+
+private:
+    std::function<void()> m_callback;
+    std::thread::id m_id;
```

Comment:
Shouldn't this now be initialized in the constructor? Or are the default constructor actually initializing it from the current thread id?

Reply:
it is now https://github.com/realm/realm-kotlin/pull/448/commits/4efc56baf3eaab12bc3e014c68f3ddbb48538a4a#diff-c71ff0791782da90014db6f7000bdeea6c883ca9729f17bd1b11041d1e7711faR146-R148 I forgot to upload some commits